File: codeforces/1900~1999/1919/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

struct Node {
  int x, l, r;
  bool deleted;
};

void solve() {
  int n = 0;
  cin >> n;
  vector<int> a(n);
  for (int i = 0; i < n; ++i) cin >> a[i];

  vector<Node> nodes;
  for (int i = 0; i < n; ++i) {
    nodes.push_back(Node{a[i], i-1, i+1, false});
  }

  auto should_remove = [&](int i) -> bool {
    if (i < 0 || i >= n) return false;

    int l = nodes[i].l;
    int r = nodes[i].r;
    int x = nodes[i].x;
    return (l >= 0 && nodes[l].x == x-1) || (r < n && nodes[r].x == x-1);
  };

  priority_queue<array<int,2>, vector<array<int,2>>, less<>> pq;
  for (int i = 0; i < n; ++i) {
    if (should_remove(i)) {
      pq.push({nodes[i].x, i});
    }
  }

  while (!pq.empty()) {
    auto [_, idx] = pq.top();
    pq.pop();

    if (nodes[idx].deleted) {
      continue;
    }

    nodes[idx].deleted = true;
    int l = nodes[idx].l;
    int r = nodes[idx].r;
    if (l >= 0) nodes[l].r = r;
    if (r < n) nodes[r].l = l;

    if (should_remove(l)) {
      pq.push({nodes[l].x, l});
    }
    if (should_remove(r))  {
      pq.push({nodes[r].x, r});
    }
  }

  int min_val = numeric_limits<int>::max();
  int left = 0;
  for (int i = 0; i < n; ++i) if (!nodes[i].deleted) {
    min_val = min(min_val, nodes[i].x);
    left += 1;
  }

  if (left == 1 && min_val == 0) {
    cout << "YES" << endl;
  } else {
    cout << "NO" << endl;
  }
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1900~1999/1919/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BinaryHeap;
use std::cmp::min;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

struct Node {
    x: i32,
    l: i32,
    r: i32,
    deleted: bool,
}

fn should_remove(i: i32, nodes: &Vec<Node>) -> bool {
    let n = nodes.len();
    if i < 0 || i >= n as i32 {
        return false;
    }
    let i = i as usize;
    let l = nodes[i].l;
    let r = nodes[i].r;
    let x = nodes[i].x;
    (l >= 0 && nodes[l as usize].x == x-1) || (r < n as i32 && nodes[r as usize].x == x-1)
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let n = read_int::<usize>();
    let a = read_vec::<i32>();
    let mut nodes: Vec<Node> = Vec::new();
    for i in 0..n {
        nodes.push(Node{
            x: a[i],
            l: i as i32 - 1,
            r: i as i32 + 1,
            deleted: false,
        });
    }


    let mut pq: BinaryHeap<(i32,i32)> = BinaryHeap::new();
    for i in 0..n {
        if should_remove(i as i32, &nodes) {
            pq.push((nodes[i].x, i as i32));
        }
    }

    while !pq.is_empty() {
        if let Some(item) = pq.pop() {
            let idx = item.1 as usize;

            if nodes[idx].deleted {
                continue;
            }
            
            nodes[idx].deleted = true;
            let l = nodes[idx].l;
            let r = nodes[idx].r;
            if l >= 0 {
                nodes[l as usize].r = r;
            }
            if r < n as i32 {
                nodes[r as usize].l = l;
            }

            if should_remove(l, &nodes) {
                pq.push((nodes[l as usize].x, l));
            }
            if should_remove(r, &nodes) {
                pq.push((nodes[r as usize].x, r));
            }
        }
    }

    let mut left = 0;
    let mut min_val = i32::MAX;
    for i in 0..n {
        if !nodes[i].deleted {
            left += 1;
            min_val = min(min_val, nodes[i].x);
        }
    }

    if left == 1 && min_val == 0 {
        writeln!(writer, "YES").ok();
    } else {
        writeln!(writer, "NO").ok();
    }
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
